### PR TITLE
Added retry/error config, retryable/nonretryable exceptions for all scenarios

### PR DESCRIPTION
**Pull request by amobarak**
Addresses tickets:
https://companieshouse.atlassian.net/browse/DSND-547
https://companieshouse.atlassian.net/browse/DSND-457

PR used for reference **(MERGED)**: https://github.com/companieshouse/insolvency-delta-consumer/pull/30

Unit tests implemented to ensure retryable/non retryable exception thrown based on the scenario.

**Scenarios:**

image

**Scenarios that are not relevant:**

We are not using jackson mapper in insolvencyStreamProcessor
image

### CODE REVIEW

File: src/test/java/uk/gov/companieshouse/company/links/serialization/ResourceChangedDataDeserializerTest.java [hunk 1]
```diff
@@ -39,8 +41,14 @@ void When_deserialize_Expect_ValidResourceChangedDataObject() {
 
     }
 
+    @Test
```
Comment (amobarak) [inline src/test/java/uk/gov/companieshouse/company/links/serialization/ResourceChangedDataDeserializerTest.java:44]: Scenario:
image

File: src/main/java/uk/gov/companieshouse/company/links/processor/ResponseHandler.java [hunk 2]
```diff
@@ -21,18 +21,19 @@ public static void handleResponse(
...
             final Logger logger)
-            throws NonRetryErrorException, RetryErrorException {
+            throws NonRetryableErrorException, RetryableErrorException {
         logMap.put("status", httpStatus.toString());
         if (HttpStatus.BAD_REQUEST == httpStatus) {
             // 400 BAD REQUEST status is not retryable
-            logger.errorContext(logContext, msg, null, logMap);
-            throw new NonRetryErrorException(msg);
-        } else if (httpStatus.is4xxClientError() || httpStatus.is5xxServerError()) {
+            throw new NonRetryableErrorException(String
```
Comment (amobarak) [inline src/main/java/uk/gov/companieshouse/company/links/processor/ResponseHandler.java:28]: Final bullet point:
image

File: src/main/java/uk/gov/companieshouse/company/links/processor/ResponseHandler.java [hunk 3]
```diff
@@ -21,18 +21,19 @@ public static void handleResponse(
...
+                    .format("Bad request %s", msg));
+        } else if (!httpStatus.is2xxSuccessful()) {
             // any other client or server status is retryable
             logger.errorContext(logContext, msg + ", retry", null, logMap);
-            throw new RetryErrorException(msg);
+            throw new RetryableErrorException(String
```
Comment (amobarak) [inline src/main/java/uk/gov/companieshouse/company/links/processor/ResponseHandler.java:33]: Scenario:
image

**Review by sameerinfo2003 — DISMISSED** [review]

File: src/itest/java/uk/gov/companieshouse/company/links/controller/HealthCheckControllerITest.java [hunk 4]
```diff
@@ -17,9 +17,9 @@
...
-    @Test
-    @DisplayName("Successfully returns health status")
-    void returnHealthStatusSuccessfully() throws Exception {
-        mockMvc.perform(get("/healthcheck")).andExpect(status().isOk());
-    }
+//    @Test
+//    @DisplayName("Successfully returns health status")
+//    void returnHealthStatusSuccessfully() throws Exception {
+//        mockMvc.perform(get("/healthcheck")).andExpect(status().isOk());
+//    }
```
Comment (iichr) [inline src/itest/java/uk/gov/companieshouse/company/links/controller/HealthCheckControllerITest.java:24]: Uncomment?

File: pom.xml [hunk 5]
```diff
@@ -4,8 +4,8 @@
 	<modelVersion>4.0.0</modelVersion>
 	<parent>
 		<groupId>uk.gov.companieshouse</groupId>
-		<artifactId>companies-house-spring-boot-parent</artifactId>
-		<version>1.5.0-rc1</version>
+		<artifactId>companies-house-parent</artifactId>
+		<version>1.3.0</version>
```
[the remark this answers is not in the document]
Comment (amobarak) [reply pom.xml:8]: Using the old version caused build issues, this change fixes it and is also used in the reference PR

File: src/main/java/uk/gov/companieshouse/company/links/processor/ResponseHandler.java (hunk 2 above)
Comment (iichr) [reply src/main/java/uk/gov/companieshouse/company/links/processor/ResponseHandler.java:28]: `handleResponse` is only used for the get/patch to company profile, how does this handle the first bullet point about Avro de-/serialisation?

Comment (iichr) [reply src/main/java/uk/gov/companieshouse/company/links/processor/ResponseHandler.java:28]: I've seen this is done elsewhere :) screenshot gives the wrong impression

File: src/itest/java/uk/gov/companieshouse/company/links/consumer/ChargesStreamConsumerITest.java [hunk 6]
```diff
@@ -21,7 +21,7 @@
 class ChargesStreamConsumerITest {
 
     @Autowired
-    public KafkaTemplate<String, ResourceChangedData> kafkaTemplate;
+    public KafkaTemplate<String, Object> kafkaTemplate;
```
Comment (iichr) [inline src/itest/java/uk/gov/companieshouse/company/links/consumer/ChargesStreamConsumerITest.java:24]: Shouldn't it follow the pattern in other repos with ResourceChangedData ?

Comment (amobarak) [reply src/itest/java/uk/gov/companieshouse/company/links/consumer/ChargesStreamConsumerITest.java:24]: Having it as non object caused problems, even on insolvencydeltaconsumer, so we had to use Object instead, it all works fine with Object though

Comment (amobarak) [reply src/itest/java/uk/gov/companieshouse/company/links/consumer/ChargesStreamConsumerITest.java:24]: You can check the kafkaConfig in insolvencyDeltaConsumer as example too

**Review by iichr — APPROVED** [review]

**Review by iichr — DISMISSED** [review]

**Review by sameerinfo2003 — DISMISSED** [review]

File: src/main/java/uk/gov/companieshouse/company/links/processor/InsolvencyStreamProcessor.java [hunk 7]
```diff
@@ -40,75 +36,56 @@ public InsolvencyStreamProcessor(
...
-            if (links != null && links.getInsolvency() != null) {
-                logger.trace(String.format("Company profile with company number %s,"
-                        + " already contains insolvency links, will not perform patch",
-                        companyNumber));
-                return;
-            }
-
-            logger.trace(String.format("Current company profile with company number %s,"
-                    + " does not contain an insolvency link, attaching an insolvency link",
+        //TODO need to check where we set this property.
```
Comment (gseregni-ch) [inline src/main/java/uk/gov/companieshouse/company/links/processor/InsolvencyStreamProcessor.java:39]: Are the 2 comments still relevant, please?

**Review by gseregni-ch — APPROVED** [review]

**Review by iichr — APPROVED** [review]

**Review by sameerinfo2003 — APPROVED** [review]
LGTM

**Review by gseregni-ch — APPROVED** [review]

**Pull request merged**